### PR TITLE
Switch to Github Actions pytests for Windows and Ubuntu

### PR DESCRIPTION
**Pull request by jwj019**

### CODE REVIEW

File: .github/workflows/windows-test.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+          Data
+        key: notebook-data-{{ checksum "Data/manifest.json" }}
+        restore-keys: |
+          notebook-data-{{ checksum "Data/manifest.json" }}
+          notebook-data-
+    - name: Debug path
+      run: ls
+    - name: Debug data path
+      run: ls Data
+    - name: Set up Python ${{ matrix.python-version }}
```
Comment (zivy) [inline .github/workflows/windows-test.yml:30]: Remove the "Debug path" and "Debug data path" steps if not needed in the final script.

File: Python/71_Trust_But_Verify.ipynb [hunk 2]
```diff
@@ -80,7 +80,7 @@
     "%run update_path_to_download_script\n",
     "from downloaddata import fetch_data as fdata\n",
     "\n",
-    "OUTPUT_DIR = 'Output'\n",
+    "OUTPUT_DIR = './Output'\n",
```
Comment (zivy) [inline Python/71_Trust_But_Verify.ipynb:83]: Why was this modified?

File: Python/71_Trust_But_Verify.ipynb [hunk 3]
```diff
@@ -532,8 +532,8 @@
     "        with tempfile.TemporaryDirectory() as tmpdirname:\n",
     "            if platform.system() == 'Windows':\n",
     "                for i, fname in enumerate(image_file_name):\n",
-    "                    shutil.copy(os.path.abspath(fname),\n",
-    "                               os.path.join(tmpdirname,str(i)))\n",
+    "                    os.symlink(os.path.abspath(fname),\n",
+    "                               os.path.join(tmpdirname,str(i)))  \n",
     "            else:\n",
```
Comment (zivy) [inline Python/71_Trust_But_Verify.ipynb:537]: Why was the shutil.copy changed to os.symlink? On windows 10 for a standard user to create symbolic links requires an update to the OS settings or admin privileges (see os.symlink [documentation](https://docs.python.org/3/library/os.html)).

File: requirements_dev.txt [hunk 4]
```diff
@@ -3,4 +3,5 @@ markdown
 lxml
 pyenchant
 ipython
+matplotlib
```
Comment (zivy) [inline requirements_dev.txt:6]: No need for matplotlib in the requirements_dev.txt as it is included through the requirements.txt.

File: tests/additional_dictionary.txt [hunk 5]
```diff
@@ -5,6 +5,7 @@ AddTransform
 Affine
 AffineTransform
 Another's
```
Comment (zivy) [inline tests/additional_dictionary.txt:7]: What is the reason for adding additional words to the custom dictionary?

File: tests/test_notebooks.py [hunk 6]
```diff
@@ -269,13 +271,13 @@ def dynamic_analysis(self, path, kernel_name):
 
         # Execute the notebook and allow errors (run all cells), output is
         # written to a temporary file which is automatically deleted.
-        with tempfile.NamedTemporaryFile(suffix='.ipynb') as fout:
+        with tempfile.NamedTemporaryFile(suffix='.ipynb',mode='wb+',delete=False) as fout:
             args = ['jupyter', 'nbconvert',
```
Comment (zivy) [inline tests/test_notebooks.py:275]: What was the motivation for this change? Mode is same as default and we do want the output file to automatically be deleted after it is analyzed.

File: tests/test_notebooks.py [hunk 7]
```diff
@@ -201,7 +202,8 @@ def static_analysis(self, path):
               for document_link in html_tree.iterlinks():
                  try:
                     if all( prefix not in document_link[2] for prefix in ['http','https']):  # Local file.
-                       url = 'file://' + os.path.abspath(document_link[2])
+                     #   formatted_path = os.path.abspath()
```
Comment (zivy) [inline tests/test_notebooks.py:205]: Delete commented out code.

**Review by zivy — CHANGES_REQUESTED** [review]

File: tests/additional_dictionary.txt (hunk 5 above)
Comment (jwj019) [reply tests/additional_dictionary.txt:7]: As far as I could tell, these words were not included in the words in the dictionaries Windows used so spelling errors were getting flagged.  I couldn't find a way to install other dictionaries with the Windows version of pyenchant.  The only workaround I could find was to add these words in the custom dictionary.

File: .github/workflows/patch-asyncio.ps1 [hunk 8]
```diff
@@ -0,0 +1,3 @@
+$fileName = 'c:\hostedtoolcache\windows\python\3.8.5\x64\lib\asyncio\__init__.py'
```
Comment (blowekamp) [inline .github/workflows/patch-asyncio.ps1:1]: This patch looks like a bit of a hack. It will only work with this specific version of python. When the GHA machine image is updated to a newer version it will break. Please document this hack a little better, is there a Python issue to reference or something?

File: .github/workflows/ubuntu-test.yml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - uses: actions/cache@v2
+      with:
+        path: |
+          Data
+        key: notebook-data-{{ checksum "Data/manifest.json" }}
+        restore-keys: |
+          notebook-data-{{ checksum "Data/manifest.json" }}
+          notebook-data-
```
Comment (blowekamp) [inline .github/workflows/ubuntu-test.yml:25]: Nice cache usage!

**Comment by blowekamp** [comment]
Nice work! 👍 

The CI config files look very good.

Please consider squashing the commit history into fewer logical commits. Perhaps with message which address some of Ziv's questions.

**Comment by blowekamp** [comment]
Also I have never figure out how to get GHA to run on a PR with adds the actions. @jwj019 Do you know if this can be done?

File: .github/workflows/patch-asyncio.ps1 (hunk 8 above)
Comment (jwj019) [reply .github/workflows/patch-asyncio.ps1:1]: Yeah, it's admittedly pretty hacky. The issue comes from the Tornado web server that Jupyter relies on.  The issue is documented at https://github.com/tornadoweb/tornado/issues/2608.  It's still current as seen at https://www.tornadoweb.org/en/stable/releases/v6.0.4.html#general-changes.  There's an upstream issue in the asyncio library that can be seen at https://bugs.python.org/issue37373.

To address this patch breaking when Python gets upgraded, I can apply the patch in a virtualenv instead which wouldn't require a hard path with a version in it. Would that work?

File: tests/test_notebooks.py (hunk 6 above)
Comment (jwj019) [reply tests/test_notebooks.py:274]: I'll remove the mode here.  I had different modes explicitly stated while I was troubleshooting an issue here.  

The delete=False is here to address an issue in Windows that seems to have with temporary files (see https://bugs.python.org/issue14243).  Would just deleting the file after we're done with everything be sufficient?

File: Python/71_Trust_But_Verify.ipynb (hunk 2 above)
Comment (jwj019) [reply Python/71_Trust_But_Verify.ipynb:83]: From what I could tell, Windows was handling the path without ./ as an absolute.  Adding the ./ forced it to be relative.

File: Python/71_Trust_But_Verify.ipynb (hunk 3 above)
Comment (jwj019) [reply Python/71_Trust_But_Verify.ipynb:537]: This was a workaround for an hour I was getting in Windows, but I can't seem to remember what the error was at the moment.  I'll go back and retest this.

File: tests/test_notebooks.py (hunk 6 above)
Comment (zivy) [reply tests/test_notebooks.py:274]: Got it. Please leave the `delete=False`, and modify the comment above it to explain that the setting is due to issues with windows, possibly include the link to the bug. 

Hopefully it will be fixed in the near future and we can revert to the original settings. Given that this is an eight year old issue I don't expect much, but then again it recently received some attention. Hope springs eternal.

File: Python/71_Trust_But_Verify.ipynb (hunk 2 above)
Comment (zivy) [reply Python/71_Trust_But_Verify.ipynb:83]: I don't think this is the issue, as the same name is used in all the notebooks and I assume those passed the testing on windows. Not sure what is different in this notebook.

Comment (jwj019) [reply Python/71_Trust_But_Verify.ipynb:84]: Ah, you're right. I just retested this, and it works fine in the original form.  I must have gotten it mixed it up with other issues.  I'll revert this.

File: .github/workflows/patch-asyncio.ps1 (hunk 8 above)
Comment (blowekamp) [reply .github/workflows/patch-asyncio.ps1:1]: You could obtain the path with something like `python -c "import asyncio; print(asyncio.__file__)"`, but then it may be easier to write the whole script in python.

>To address this patch breaking when Python gets upgraded, I can apply the patch in a virtualenv instead which wouldn't require a hard path with a version in it. Would that work?

That could work two, but I does not look like a virtual env is being used yet.

File: Python/71_Trust_But_Verify.ipynb (hunk 3 above)
Comment (jwj019) [reply Python/71_Trust_But_Verify.ipynb:538]: This seems to work fine with shutil.copy.  Not sure what the original error was. Thanks for pointing this out.

**Comment by jwj019** [comment]
Ok, I think I've taken care of all the remaining issues and squashed it all down into 2 commits.  @blowekamp @blowekamp Please give your final review when you get a chance.

**Review by zivy — COMMENTED** [review]
Please rebase master (some changes already merged there - registration_utilities.py).
Repository only has master branch, no release, so that can be removed from the yml branch specifications.

**Review by blowekamp — APPROVED** [review]
The code looks very good to me.

It is our convention to prefer rebasing onto master over merging master into a topic branch or PR. It just reduces the number of merges in the final master branch to keep the history neat and tidy.

**Review by zivy — COMMENTED** [review]
Something went wrong when rebasing/merging, many strange changes and there is a merge conflict in the environment.yml file that wasn't resolved. 

I'd recommend moving all the new stuff to the side and redoing the PR; updating master from origin, creating a new branch off of the updated master, redoing the two commits and then opening a PR.

**Comment by jwj019** [comment]
Yep, closing this PR and taking up your recommendations
